compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/IsolationAnalyzer.java

```diff
@@ -3213,7 +3213,7 @@ private boolean isEnclosedLockWithinSameFunction(BLangNode parent, BLangLock pot
             return true;
         }
 
-        if (parent == null || parent.getKind() == NodeKind.FUNCTION) {
+        if (parent == null || parent.getKind() == NodeKind.FUNCTION || parent.getKind() == NodeKind.RESOURCE_FUNC) {
```

Comment:
Created https://github.com/ballerina-platform/ballerina-lang/issues/37399 to verify whether this check is required.